Simplify \exists with bound substitutions

ttuegel opened the pull request:
###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`
- [ ] Style conformance: `stylish-haskell`

---

virgil-serbanuta commented on kore/src/Kore/Step/Simplification/Exists.hs:
```diff
@@ -236,45 +235,50 @@ makeEvaluateNoFreeVarInSubstitution
-                        )
-                , predicate = makeTruePredicate
-                , substitution = substitution
+                    $ Maybe.fromMaybe (mkTop patternSort)
```
I know that this `mkTop patternSort` is not actually used, but, since it's here, locally it looks weird that we're adding an `exists` on something that might be a `top`. Fmapping the mkExists before calling `fromMaybe` might look better.

virgil-serbanuta commented on kore/src/Kore/Step/Simplification/Exists.hs:
```diff
@@ -236,45 +235,50 @@ makeEvaluateNoFreeVarInSubstitution
+                , predicate =
+                    Predicate.makeAndPredicate freePredicate
+                    $ Predicate.makeExistsPredicate variable
+                    $ Maybe.fromMaybe Predicate.makeTruePredicate boundCondition
```
The same as above: why add a quantifier on a `top` value?

ttuegel commented on kore/src/Kore/Step/Simplification/Exists.hs:
```diff
@@ -198,16 +199,16 @@ makeEvaluate
-                    localSubstitutionList
-                    (Substitution.wrap globalSubstitution)
+                    (Map.singleton variable boundTerm)
+                    (Substitution.unsafeWrap $ Map.toList freeSubstitution)
```
This step will only work if the substitution is normalized, otherwise we can't apply the substitution correctly. Should we call `normalize` here to ensure that it is normalized, or fall back to the second case if it is not?

virgil-serbanuta replied:
I'm worried that there is no simple and obvious way (preferably one that's also stable through code changes) to see that `freeSubstitution` is, indeed, normalized.

Calling `normalize` here should be fine. Using `Kore.Unification.Substitution.partition` or something close to it (but declared in the same module) that produces a wrapped substitution would also be fine.

ttuegel commented:
@virgil-serbanuta This is ready for your re-review. I changed `Kore.Step.Simplification.Exists.makeEvaluate` to normalized the substitution before doing anything that depends on the normalization state. I also discovered that we were doing some things with `Kore.Unification.Substitution.toMap` which were not quite safe; I changed a few types and added some error cases to catch uses of `toMap` on denormalized substitutions.

virgil-serbanuta commented on kore/src/Kore/Unification/Substitution.hs:
```diff
@@ -35,71 +38,108 @@ import           Data.Set
+    -- TODO (thomas.tuegel): Instead of a sum type, use a product containing the
+    -- normalized and denormalized parts of the substitution together. That
+    -- would enable us to keep more substitutions normalized in the Semigroup
+    -- instance below.
```
Are you sure it's worth doing that? What would be the benefit? And it might not be trivial - consider `x=f(y)`: it's ok to merge it with `z=g(y)`, but not with any of `x=g(z)`, `y=g(z)`, `z=g(x)`, or `z=sigma(y) /\ z = sigma(g(t))`.

ttuegel replied:
I think it's worth it if we keep substitution deduplication as a separate step, because we can keep the substitution normalized (well, semi-normalized: cycle-free) incrementally. The most expensive thing we do during execution is merging and re-normalizing substitutions; it's expensive in large part because merging/re-normalization always begins by assuming nothing is normalized. I have in mind a representation such as `data Substitution = Substitution { normal = Map variable (MultiAnd pattern), denormal = Map variable (MultiAnd pattern) }` with the invariant that no variable occurs as a key in both `normal` and `denormal` and the keys of `normal` do not occur in any values of `normal` or `denormal`.

I would merge those examples in the following ways, where each substitution is represented as a pair as described above:

1. `([x=f(y)], []) ∧ ([x=g(z)], [])` = `([x=f(y)∧g(z)], [])` (postponing de-duplication)
1. `([x=f(y)], []) ∧ ([y=g(z)], [])` = `([x=f(g(z)), y=g(z)], [])` (applying substitution, which has no cycles)
1. `([x=f(y)], []) ∧ ([z=g(x)], [])` = `([x=f(y), z=g(f(y))], [])` (as above)
1. `([z=σ(y)], []) ∧ ([z=σ(g(t))], [])` = ([z=σ(y)∧σ(g(t))], [])` (postponing de-duplication, as above)
1. An interesting case not mentioned: `([x=f(y), []) ∧ ([y=g(x)], [])` = `([x=f(y), y=g(f(y))])` (breaking a cycle by applying the `normal` substitution and denormalizing; there is an choice about which way we break the cycle)

Anyway, we can discuss this in more detail when performance becomes an issue, and we can decide to if and how to do this, or not. For now, I mainly wanted to document that we have the option of doing this.

virgil-serbanuta commented on kore/src/Kore/Step/Simplification/Exists.hs:
```diff
@@ -196,194 +195,199 @@ makeEvaluate
-@
-φ ∧ (freeP ∧ (∃ y. boundP ∧ boundS)) ∧ freeS.
-@
+See also: 'quantifyExpandedPattern'
```
This function relies on `boundTerm` not containing `variable` as a free variable, it may be worth adding that constraint here (another way would be to use a normalized substitution, but that may not be worth the effort).

ttuegel replied:
I documented this fact and added a check.
